### PR TITLE
Crosschasm c5 updates for Cellular device and iOS. Also requires openxc-message-format update.

### PR DESCRIPTION
**Pull request by emarsman**

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -25,7 +25,7 @@
 	url = https://code.google.com/p/nanopb/
 [submodule "src/libs/openxc-message-format"]
 	path = src/libs/openxc-message-format
-	url = https://github.com/openxc/openxc-message-format
+	url = https://github.com/mgiannikouris/openxc-message-format
```
Comment (peplin) [inline .gitmodules]: Could you change this back to the `openxc` source?

File: CELLULAR.mkd [hunk 2]
```diff
@@ -0,0 +1,89 @@
+## CONFIGURATION OPTIONS SPECIFIC TO THE CELLULAR C5
```
Comment (peplin) [inline CELLULAR.mkd]: These docs should go in the `docs` directory and be added to one of the indexes, maybe from the advanced configuration options page - that way it will be available at vi-firmware.openxcplatform.com

File: fabfile.py [hunk 3]
```diff
@@ -27,7 +27,8 @@
 env.boards = {
     "reference": {"name": "FORDBOARD", "extension": "bin"},
     "chipkit": {"name": "CHIPKIT", "extension": "hex"},
-    "c5": {"name": "CROSSCHASM_C5", "extension": "hex"}
+    "c5": {"name": "CROSSCHASM_C5", "extension": "hex"},
+	"cellc5": {"name": "CROSSCHASM_CELLULAR_C5", "extension": "hex"}
```
Comment (peplin) [inline fabfile.py]: The indentation is off here

File: src/Makefile [hunk 4]
```diff
@@ -153,6 +155,14 @@ ifeq ($(PLATFORM), CHIPKIT)
 include platform/pic32/pic32.mk
 else ifeq ($(PLATFORM), CROSSCHASM_C5)
 include platform/pic32/pic32.mk
+else ifeq ($(PLATFORM), CROSSCHASM_CELLULAR_C5)
+# force a modified set of linkers for cellular c5
+# this is necessary to support OTA upgrades
```
Comment (peplin) [inline src/Makefile]: OTA? Niiiiice

**Comment by peplin** [comment]
Globally, can you replace tabs with 4 spaces to match the rest of the code?

File: src/config.cpp [hunk 5]
```diff
@@ -1,29 +1,37 @@
...
+	getFlashHash(config);
+	config->telit.uart = &config->uart;
     config->initialized = true;
 }
 
 openxc::config::Configuration* openxc::config::getConfiguration() {
     static openxc::config::Configuration CONFIG = {
         messageSetIndex: 0,
-        version: "7.0.1-dev",
+        version: "7.0.1-c5-dev",
```
Comment (peplin) [inline src/config.cpp]: You probably didn't mean to leave this in here - we could make it more dynamic, so the version is set based on the platform.

File: src/payload/json.cpp [hunk 6]
```diff
@@ -564,6 +596,8 @@ int openxc::payload::json::serialize(openxc_VehicleMessage* message,
     size_t finalLength = 0;
     if(root != NULL) {
         bool status = true;
+		if(message->has_uptime)
```
Comment (peplin) [inline src/payload/json.cpp]: A small thing, but the code base always uses curly braces even for 1 line conditionals to avoid errors in the future.

File: src/pipeline.cpp [hunk 7]
```diff
@@ -137,7 +155,11 @@ void openxc::pipeline::publish(openxc_VehicleMessage* message,
 void openxc::pipeline::sendMessage(Pipeline* pipeline, uint8_t* message,
         int messageSize, MessageClass messageClass) {
     sendToUsb(pipeline, message, messageSize, messageClass);
-    sendToUart(pipeline, message, messageSize, messageClass);
+	#ifdef TELIT_HE910_SUPPORT
```
Comment (peplin) [inline src/pipeline.cpp]: Why the either/or connection?

File: src/platform/pic32/uart.cpp [hunk 8]
```diff
@@ -26,6 +26,10 @@
 #include "atcommander.h"
 #include "WProgram.h"
 #include "gpio.h"
+#include "util/timer.h"
+
+#warning "mikeG add some plib"
```
Comment (peplin) [inline src/platform/pic32/uart.cpp]: What's up here?

File: src/platform/pic32/uart.cpp [hunk 9]
```diff
@@ -83,6 +94,8 @@ void openxc::interface::uart::changeBaudRate(UartDevice* device, int baud) {
             baud) - 1);
     ((p32_uart*)_UART1_BASE_ADDRESS)->uxMode.reg = (1 << _UARTMODE_ON) |
             (1 << _UARTMODE_BRGH);
+			#warning "MG UEN"
```
Comment (peplin) [inline src/platform/pic32/uart.cpp]: What does UEN set? What's the warning for?

File: src/platform/pic32/uart.cpp [hunk 10]
```diff
@@ -163,7 +181,32 @@ bool openxc::interface::uart::connected(UartDevice* device) {
             status = false;
             break;
     }
-    return status;
 
 #endif
+
+	// on apple devices, it takes some time for flow control to propagate to the device
```
Comment (peplin) [inline src/platform/pic32/uart.cpp]: Very interesting - is this the source of the bug where it would disconnect after ~5 seconds?

File: src/platform/pic32/uart.cpp [hunk 11]
```diff
@@ -163,7 +181,32 @@ bool openxc::interface::uart::connected(UartDevice* device) {
...
+	// during that time, if we transmit, we crash the connection
+	// simple solution: rising delay for device connected status
+	if(last_status == false && status == true)
```
Comment (peplin) [inline src/platform/pic32/uart.cpp]: The braces here don't match the style of the rest of the code - start on the same line as the conditional.

File: src/telit_he910.cpp [hunk 12]
```diff
@@ -0,0 +1,2102 @@
...
+	static bool pdp_connected = false;
+	static uint8_t pdp_counter = 0;
+
+	// device stats
+	static unsigned int SIMstatus = 0;
+	static NetworkConnectionStatus connectStatus = UNKNOWN;
+	static char ICCID[32];
+	static char IMEI[32];
+	
+	switch(state)
```
Comment (peplin) [inline src/telit_he910.cpp]: Is it possible to pull up the handlers for each state into a function? This switch is pretty big and hard to follow.

File: src/telit_he910.cpp [hunk 13]
```diff
@@ -0,0 +1,2102 @@
...
+		
+			state = POWER_OFF;
+		
+			break;
+	}
+	
+	return state;
+}
+
+void openxc::telitHE910::deinitialize() {
```
Comment (peplin) [inline src/telit_he910.cpp]: This brace style is right, but elsewhere in this file it's not matching the code base.

File: src/pipeline.cpp (hunk 7 above)
Comment (peplin) [inline src/pipeline.cpp]: I see now as I read through, it's taking over the UART communication to talk to the modem. I feel like that could be more clear in the code, but I'm not sure how.

File: src/telit_he910.cpp [hunk 14]
```diff
@@ -0,0 +1,2102 @@
...
+	return rc;
+
+}
+
+bool openxc::telitHE910::setBaud(unsigned int baudRate) {
+
+	bool rc = true;
+	char command[32] = {};
+	
+	sprintf(command, "AT+IPR=%u\r\n", baudRate);
```
Comment (peplin) [inline src/telit_he910.cpp]: If this uses AT-style commands, I wonder if you could build support for it into https://github.com/openxc/at-commander since we already use that to control the RN-4x module for Bluetooth.

File: src/telit_he910.cpp [hunk 15]
```diff
@@ -0,0 +1,2102 @@
...
+	while(!QUEUE_EMPTY(uint8_t, &device->sendQueue) && (pSendBuffer - sendBuffer) < sendBufferSize)
+	{
+        *pSendBuffer++ = QUEUE_POP(uint8_t, &device->sendQueue);
+    }
+
+	return;
+
+}
+
+#endif
```
Comment (peplin) [inline src/telit_he910.cpp]: This file is very large - can you make it a subfolder and break it up into some logical chunks?

**Comment by peplin** [comment]
Overall - awesome! There's some minor style things, but it's great how neatly this fit into the architecture used in the firmware already.

One concern - there are no unit tests for any of the new code. I'd like to at least see a few basic tests to see that the modem functions are called when it is enabled.

File: .gitmodules (hunk 1 above)
Comment (emarsman) [inline .gitmodules]: I issued a pull request for the message format as well. The new version doesn't exist yet in openxc. Do you have a preference for the sequence of these with the dependency?

File: src/config.cpp (hunk 5 above)
Comment (emarsman) [inline src/config.cpp]: Updated to 7.0.2-dev

File: fabfile.py (hunk 3 above)
Comment (emarsman) [inline fabfile.py]: Fixed.

File: src/payload/json.cpp (hunk 6 above)
Comment (emarsman) [inline src/payload/json.cpp]: Fixed globally.

File: src/platform/pic32/uart.cpp (hunk 11 above)
Comment (emarsman) [inline src/platform/pic32/uart.cpp]: Fixed.

File: src/telit_he910.cpp (hunk 13 above)
Comment (emarsman) [inline src/telit_he910.cpp]: Fixed.

**Comment by emarsman** [comment]
I'll have to talk to @mgiannikouris about some of the larger issues that are still unresolved. 

@peplin Do you have docs or recommendations for adding the tests? Also, info on how to run the travis builds locally before doing a push? Is that the recommended way for CI?

**Comment by peplin** [comment]
Hey Eric, yes there's a couple of quick docs here on the test suite: http://vi-firmware.openxcplatform.com/en/master/testing.html#test-suite

To summarize, each supported platform consolidates its board-specific code under the [`platforms` directory](https://github.com/openxc/vi-firmware/tree/master/src/platform). Thest of the code base is just regular ol' C (with some C++ nicities used sparingly). We also [implement those platform-specific functions with test stubs](https://github.com/openxc/vi-firmware/tree/master/src/tests/platform) so we can compile everything on a regular x86 architecture with gcc or clang. There are unit and functional tests in the `tests` directory, and then recently I added a complete integration test suite that runs on the actual hardware. That level of testing is probably practical for this case, since it would involve an external server.

File: src/platform/pic32/uart.cpp (hunk 8 above)
Comment (mgiannikouris) [inline src/platform/pic32/uart.cpp]: Shouldn't have been left behind. I thought I needed to include plib to get access to the UART registers but I did not. Removing on next commit.

File: src/platform/pic32/uart.cpp (hunk 10 above)
Comment (mgiannikouris) [inline src/platform/pic32/uart.cpp]: I believe so, yes. I monitored the UART lines with a scope during the process of connecting and sending data. Basically the "connected" indicator line on the RN42 goes high as soon as the devices connects to an iOS device. However, the application that handles the incoming iAP data is not necessary open at this time. If it is not, the flow control signal is supposed to go high (if we transmit too many bytes while there is no app open to handle our data, the link eventually breaks). However there appears to be a rather significant (and not consistent) delay in the flow control line propagating.

The current solution of a delay seems to be a simple and effective resolution given the current hardware and software setup, although it's inefficient. There is a more proper way I think, which involves an RN42 GPIO that goes high when an app has opened (which would replace the "connected" indicator line, although how this plays with non-iOS devices I'm not sure). There's also a set of diagnostic messages that can be enabled to indicate SESSION_OPEN and SESSION_CLOSED, but they are unsolicited so the RN42 control code needs to be able to handle that.

File: src/platform/pic32/uart.cpp (hunk 9 above)
Comment (mgiannikouris) [inline src/platform/pic32/uart.cpp]: Should have taken the warning out. The subsequent line U1MODEbits.UEN = 2; was added because uxMode register is re-initialized during the call to changeBaudRate, but not all of the potentially relevant settings within the register (i.e. peripheral flow control) that the application may have already configured are preserved. 

To be fair I don't know if this was required. I made this change while searching for the reason why iOS devices were not working, thinking that it was possible that the call to begin() might be overriding an existing flow control setting that was never restored. I didn't scan through all the ways the UART driver code is called, and depending on when changeBaudRate is called (and in relation to enableFlowControl and others) it may not matter if changeBaudRate wipes out the register, but to be safe I put it in since flow control is always used.

It might be better to use the SET, INV, and CLR registers for low-level peripheral register modification, or maybe store the register before re-initializing the peripheral, then this line can come out and the call to changeBaudRate won't have any unexpected side effects.

If I've missed the mark feel free to remove and see if iOS connectivity still works.

File: src/pipeline.cpp (hunk 7 above)
Comment (mgiannikouris) [inline src/pipeline.cpp]: Yeah you've got it. The reason this is necessary is because the existing devices (RN42) attached to the UART, once "initialized", are in "live" mode with the remote (phone) and don't execute any more device commands. Once initialized they give up the physical interface to the application code via a "connected" or "ready" flag so the application can interact with the remote.

The Telit interacts with the remote (server) via socket commands (all socket send and receive data is wrapped up in socket send and receive commands) so the device layer is needed at all times. Although a live mode is also possible with the Telit, socket commands are nice because they relax the timing requirements for the MCU. If we stop reading from a socket mid-transfer we aren't going to lose data (unless the TCP connection times out). Admittedly this comes with a potential penalty on throughput due to additional overhead.

More generally the separation of the physical interface from the application let's us give the device controller autonomy to manage the device and execute device commands as needed, independent of the application logic (e.g. find out why a connection dropped and try to restart it, while ignoring or buffering data the application is trying to send to the remote). Actually in this case the application calls the server APIs, which then call into the device controller, although I haven't separated these two properly.

File: src/telit_he910.cpp (hunk 15 above)
Comment (mgiannikouris) [inline src/telit_he910.cpp]: Agreed, I'll look at splitting this up. At minimum the server API stuff should be split off from the telit controller.

**Comment by peplin** [comment]
I did see your changes, it's looking good. Short of adding new tests, could you try and make it compile OK with the test suite we run in CI? Everything has to compile with the clang compiler (it gives very nice warnings which act as a linter for C/C++). You can see the errors from CI through the red links above in this PR ^^^.
